libraries/botbuilder-dialogs-adaptive/src/recognizers/entityRecognizers/entityRecognizer.ts

```diff
@@ -6,14 +6,122 @@
  * Licensed under the MIT License.
  */
 
-import { Entity } from 'botbuilder-core';
-import { DialogContext } from 'botbuilder-dialogs';
+import { Activity, Entity, RecognizerResult } from 'botbuilder-core';
+import { DialogContext, Recognizer } from 'botbuilder-dialogs';
+import { TextEntity } from './textEntity';
 
-export interface EntityRecognizer {
-    recognizeEntities(
+/**
+ * Entity recognizers base class.
+ */
+export class EntityRecognizer extends Recognizer {
+    /**
+     * To recognize intents and entities in a users utterance.
+     *
+     * @param {DialogContext} dialogContext Dialog Context.
+     * @param {Partial<Activity>} activity Activity.
+     * @param {object} telemetryProperties Additional properties to be logged to telemetry with event.
+     * @param {object} telemetryMetrics Additional metrics to be logged to telemetry with event.
+     * @returns {Promise<RecognizerResult>} Analysis of utterance.
+     */
+    public async recognize(
+        dialogContext: DialogContext,
+        activity: Partial<Activity>,
+        telemetryProperties?: Record<string, string>,
+        telemetryMetrics?: Record<string, number>
+    ): Promise<RecognizerResult> {
+        // Identify matched intents
+        const text = activity.text || '';
+
+        const recognizerResult: RecognizerResult = {
+            text,
+            intents: {},
+        };
+
+        if (!text) {
+            // nothing to recognize, return empty result
+            return recognizerResult;
+        }
+
+        const textEntity = new TextEntity(text);
+        textEntity['start'] = 0;
+        textEntity['end'] = text.length;
+        textEntity['score'] = 1.0;
+
+        // add entities from regexRecognizer to the entities pool
+        const entityPool: Entity[] = [textEntity];
+
+        // process entities using EntityRecognizer
+        const newEntities = await this.recognizeEntities(
+            dialogContext,
+            dialogContext.context.activity.text,
+            dialogContext.context.activity.locale,
+            entityPool
+        );
+        entityPool.push(...newEntities);
+
+        // map entityPool of Entity objects => RecognizerResult entity format
+        const entities: Record<string, unknown> = {};
+        entityPool
+            .filter((e) => e !== textEntity)
+            .forEach((entityResult: Entity) => {
+                const entityType = entityResult.type.toLowerCase();
+
+                // add value
+                let values = [];
+                if (!Object.prototype.hasOwnProperty.call(entities, entityType)) {
```

Comment:
Is there a good reason to use `hasOwnProperty` from the Object prototype? Why can't this be `if (!entities[entityType]) { ... }`

Reply:
I think this code could be more succinctly written as:

```typescript
entities[entityType] ??= [];
entities[entityType].push(...);
```

Reply:
Yes, you are right.